Adding astyle hook, and running astyle on repository

blu-base opened the pull request:
adding a `pre-commit` hook to `.githooks` directory.
The hook only runs on committed files.

This commit drops a brief sentence to the `HACKING` text file.

Anyway, many files were not fully formatted according to current specs (`astyle.options`)

blu-base committed:
Add astyle hook

blu-base committed:
Running astyle

tshikaboom commented:
Hey, that's cool! Thanks.

What do you think about checking any code, whether committed or not? This could help with a dirty workspace, i.e. code about to be committed could also be checked. Something in the vein of
```
# Get the git toplevel folder to find the files in the right place
repo_toplevel=$(git rev-parse --show-toplevel)

files=$(find $repo_toplevel -name '*.cpp' -o -name '*.hpp [other suffixes]')
```

edit: you can disregard this, I read the manpage and the `--diff-filter=ACMR` options seems to do what I'd like, and seems to be more granular than just `find`ing files in the repo.

tshikaboom commented:
Another thing I noticed when dry-running the hook is that `git-diff-index` does not appear in my `$PATH`, although invoking just `git diff-index` on my PC works.
In the case we want only committed files to be linted, could you check if that works for you too?

tshikaboom commented on .githooks/pre-commit:
```suggestion
REPO_TOPLEVEL=$(git rev-parse --show-toplevel)
```

tshikaboom commented on .githooks/pre-commit:
```suggestion
ASTYLEOPTIONS="$REPO_TOPLEVEL/astyle.options"
```
This would make it possible to use the astyle options file regardless of where we are in the repo.

tshikaboom commented on .githooks/pre-commit:
nit: there is a double tab here, instead of a single

some code around here also uses spaces :)

blu-base replied:
Oh, i missed that. It's now all spaces.

tshikaboom commented on .githooks/pre-commit:
```suggestion
files=$(git diff-index --diff-filter=ACMR --name-only -r --cached $against -- | grep -i '\.c$\|\.cpp$\|\.hpp$\|\.h$')
```

blu-base committed:
generalize git command

Co-authored-by: Oskar Viljasaar <tshikaboom@users.noreply.github.com>

blu-base committed:
generalize git command

Co-authored-by: Oskar Viljasaar <tshikaboom@users.noreply.github.com>

blu-base committed:
Update .githooks/pre-commit

Co-authored-by: Oskar Viljasaar <tshikaboom@users.noreply.github.com>

blu-base committed:
Replace tabs with spaces

blu-base commented:
Thanks for all your remarks! I didn't expecting to cause that much work.
I think all current comments and suggestions are addressed now.

tshikaboom committed:
astyle.options: do not specify the 'recursive' keyword

This makes it possible to use the config when invoking astyle with explicitly
specified filenames.

tshikaboom committed:
pre-commit: invoke astyle with the --options= argument

tshikaboom commented:
So when testing this I found that astyle rewrote all my files I committed, and the `astyle.options` file itself. (meaning it didn't use the options file, but instead used the default astyle provided options, like indent=spaces=4 etc)

Turns out it didn't like the `recursive` option in `astyle.options` when invoked explicitly with files to lint, so when manually linting everything I guess one would add a `--recursive` flag combined with the options file.

Also, it wanted the `--options=[file]` flag, that's why it linted the options file itself :)
I fixed these myself to avoid too many back and forths

tshikaboom commented:
Thanks again! It's cool to see some activity here :) and motivating me to pick up slack again!
